### ISSUE TITLE
only basetype compared for inheritance

### ISSUE THREAD

**Issue by sanderr** [inmanta/inmanta#2132]
`inmanta.ast.statements.define.DefineEntity` rejects attributes with names equal to those defined in a parent, unless the types match (line 230 at the time of writing, commit 6e9233dd). However, only the basetype is compared. This allows for models such as

```
entity A:
    number[] n
end

entity B extends A:
    number n
end

implement B using std::none

B(n = 1)
```

I believe this should produce an error. Making the type more strict could be allowed (`number?` -> `number`) but I think that is probably for a future improvement since it would require some concept of co-/contra-/invariance.

**Comment by wouterdb** [comment]
This should be resolved, the more complete inheritance can become another ticket.

We should probably get rid of the `nullable` and `multi` fields on 'Attribute',  and  construct the proper type and do a proper typecheck on that. However, this will break the compiler API. 
The cleanest way would be to introduce a new type field, with the full proper type, and mark the three old fields for deprecation (or defer the deprecation until #2129 is resolved). The associated methods could be kept (or not, depending on #2129)

This will also bring us closer to #2096 .

**Comment by sanderr** [comment]
I am currently implementing part of that to resolve inmanta/sphinx#29. I'm renaming the current `type` to `basetype` and exposing the full type in `type`. I'm leaving `multi`, `nullable` and `Attribute` construction untouched for now (`get_type` uses `basetype`, `multi` and `nullable` to determine its return value). I think doing this step by step makes it more manageable but seen as this has implications in other modules, would you prefer me to take the rest in as well while I'm at it?

**Comment by wouterdb** [comment]
I'm not entirely sure I correctly understand you, but

1. external api should remain stable ( e.g. https://github.com/inmanta/lsm/blob/428fb32fda8b1ddfc2cf4979b6665c4ad7aba475/plugins/__init__.py#L268). So renaming will break things.

2. I agree to your approach (as I understand it)
